### PR TITLE
UCX/RNDV/CUDA: RNDV protocol improvements for CUDA - v1.9.x

### PR DESCRIPTION
**Pull request by bureddy**
porting https://github.com/openucx/ucx/pull/5473 from master

### CODE REVIEW

**Review by brminich — COMMENTED** [review]
@hoopoepg, can you please take allok as well?

File: src/ucp/tag/rndv.c [hunk 1]
```diff
@@ -622,11 +582,93 @@ static void ucp_rndv_put_completion(uct_completion_t *self, ucs_status_t status)
...
+            /* Return first matching lane */
+            rndv_req->send.rndv_get.rkey_index[i] = ucs_bitmap2idx(rkey->md_map,
+                                                                   dst_md_index);
+            lane_map                             |= UCS_BIT(i);
+            max_lane_bw                           = ucs_max(max_lane_bw, lane_bw);
+        }
+    }
+
+    if (ucs_popcount(lane_map) > 1) {
+        /* remove lanes if bandwidth is too less compare to best lane */
```
Comment (hoopoepg) [inline src/ucp/tag/rndv.c:644]: remove lanes if bandwidth is too low comparing to the best lane

**Review by hoopoepg — APPROVED** [review]

File: src/ucp/tag/rndv.c [hunk 2]
```diff
@@ -1156,10 +1231,14 @@ UCS_PROFILE_FUNC_VOID(ucp_rndv_matched, (worker, rreq, rndv_rts_hdr),
                 goto out;
             }
         }
-        /* put protocol is allowed - register receive buffer memory for rma */
-        ucs_assert(rndv_rts_hdr->size <= rreq->recv.length);
-        ucp_request_recv_buffer_reg(rreq, ep_config->key.rma_bw_md_map,
-                                    rndv_rts_hdr->size);
+
+        if ((rndv_mode == UCP_RNDV_MODE_PUT_ZCOPY) ||
+            UCP_MEM_IS_CUDA(rreq->recv.mem_type)) {
```
Comment (yosefe) [inline src/ucp/tag/rndv.c:1236]: why need to check cuda here?

Comment (bureddy) [reply src/ucp/tag/rndv.c:1236]: in default UCP_RNDV_MOD_AUTO, we try to use PUT protocol for CUDA IPC case.

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
bot:pipe:retest

**Pull request merged**